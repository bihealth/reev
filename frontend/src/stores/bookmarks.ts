/**
 * Store for information regarding the current user.
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { BookmarksClient } from '@/api/bookmarks'
import { StoreState } from '@/stores/misc'

export interface BookmarkData {
  user: string
  obj_type: string
  obj_id: string
  id: string
}

export const useBookmarksStore = defineStore('bookmarks', () => {
  /* The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /* The bookmarks list for current user. */
  const bookmarks = ref<BookmarkData[]>([])

  const loadBookmarks = async () => {
    storeState.value = StoreState.Loading
    try {
      const client = new BookmarksClient()
      bookmarks.value = await client.fetchBookmarks()
      storeState.value = StoreState.Active
    } catch (e) {
      storeState.value = StoreState.Error
    }
  }

  const deleteBookmark = async (obj_type: string, obj_id: string) => {
    storeState.value = StoreState.Loading
    try {
      const client = new BookmarksClient()
      await client.deleteBookmark(obj_type, obj_id)
      await loadBookmarks()
    } catch (e) {
      storeState.value = StoreState.Error
    }
  }

  const createBookmark = async (obj_type: string, obj_id: string) => {
    storeState.value = StoreState.Loading
    try {
      const client = new BookmarksClient()
      await client.createBookmark(obj_type, obj_id)
      await loadBookmarks()
    } catch (e) {
      storeState.value = StoreState.Error
    }
  }

  const fetchBookmark = async (obj_type: string, obj_id: string) => {
    try {
      const client = new BookmarksClient()
      const response = await client.fetchBookmark(obj_type, obj_id)
      if (response.detail === 'Unauthorized') {
        storeState.value = StoreState.Error
        return null
      }
      return response
    } catch (e) {
      return null
    }
  }

  return {
    storeState,
    bookmarks,
    loadBookmarks,
    createBookmark,
    fetchBookmark,
    deleteBookmark
  }
})
