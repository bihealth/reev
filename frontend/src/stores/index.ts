export * from '@bihealth/reev-frontend-lib/stores/types'
