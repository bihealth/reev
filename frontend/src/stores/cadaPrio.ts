/**
 * Store for genes ranking from cada-prio.
 *
 * This includes the data retrieved from the APIs.
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { CadaPrioClient } from '@/api/cadaPrio'
import { StoreState } from '@/stores/misc'

export interface GeneRank {
  rank: number
  score: number
  gene_symbol: string
  ncbi_gene_id: number
  hgnc_id: string
}

/** Options for `loadData()` */
export interface LoadDataOptions {
  /** HGNC IDs to use for querying. */
  hgncIds?: string[]
  /** Whether to skip Clearing */
  skipClear?: boolean
}

export const useCadaPrioStore = defineStore('cadaPrio', () => {
  /** The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /** The current gene Ranking. */
  const geneRanking = ref<GeneRank[] | null>(null)

  function clearData() {
    storeState.value = StoreState.Initial
    geneRanking.value = null
  }

  /**
   * Run prioritization and return results for the given HPO terms and HGNC IDs.
   *
   * @param hpoTerms HPO terms to use for querying.
   * @param options Options for the query.
   */
  const loadData = async (hpoTerms: string[], options?: LoadDataOptions) => {
    // Clear against artifact unless skipped.
    if (!options?.skipClear) {
      clearData()
    } else {
      // also clear if hpoTerms is empty, then exit
      if (hpoTerms.length === 0) {
        clearData()
        return
      }
    }

    // Load data via API
    storeState.value = StoreState.Loading
    try {
      const client = new CadaPrioClient()
      const data = await client.predictGeneImpact(hpoTerms, options?.hgncIds)
      if (data.length === 0) {
        geneRanking.value = null
      } else {
        geneRanking.value = data
      }
      storeState.value = StoreState.Active
    } catch (error) {
      console.error(error)
      storeState.value = StoreState.Error
    }
  }

  return {
    storeState,
    geneRanking,
    loadData,
    clearData
  }
})
