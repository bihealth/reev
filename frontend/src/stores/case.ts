/**
 * Store for case information.
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { CaseInfoClient } from '@/api/caseinfo'
import { StoreState } from '@/stores/misc'

export enum Inheritance {
  AutosomalDominant = 'NCIT:C94245',
  AutosomalRecessive = 'NCIT:C94246',
  Cosegregation = 'NCIT:C94599',
  GeneticAnticipation = 'NCIT:C93189',
  GeneticLinkage = 'NCIT:C94542',
  XLinkedRecessive = 'NCIT:C94247',
  Unknown = 'reev:unknown_inheritance'
}

export enum Sex {
  Male = 'PATO:0000384',
  Female = 'PATO:0000383',
  Unknown = 'reev:unknown_sex'
}

export enum Ethnicity {
  AfricanAmerican = 'NCIT:C128937',
  AshkenaziJewish = 'NCIT:C17950',
  EastAsian = 'NCIT:C161419',
  Finnish = 'NCIT:C43865',
  European = 'NCIT:C43851',
  Latino = 'NCIT:C17459',
  MiddleEastern = 'NCIT:C43866',
  SouthAsian = 'NCIT:C41263',
  Other = 'NCIT:C104495',
  Unknown = 'reev:unknown_ethnicity'
}

export enum Zygosity {
  Heterozygous = 'GENO:0000135',
  Homozygous = 'NCIT:C45826',
  CompoundHeterozygous = 'NCIT:C198518',
  Unknown = 'reev:unknown_zygosity'
}

export interface OntologyTerm {
  term_id: string
  name: string
}

export interface Case {
  [key: string]: any
  /* The case pseudonym. */
  pseudonym: string
  /* Orphanet / OMIM disease(s). */
  diseases: OntologyTerm[]
  /* HPO terms. */
  hpoTerms: OntologyTerm[]
  /* Inheritance. */
  inheritance: Inheritance
  /* Affected family members. */
  affectedFamilyMembers: boolean | null
  /* Sex. */
  sex: Sex
  /* Age of onset in month. */
  ageOfOnsetMonths: number | null
  /* Ethnicity. */
  ethnicity: Ethnicity
  /* Zygosity. */
  zygosity: Zygosity
  /* Family segregation. */
  familySegregation: boolean | null
}

export interface APIResponse {
  affected_family_members: boolean
  age_of_onset_month: number
  diseases: any[] // Replace with the actual type from your API
  ethinicity: string
  family_segregation: boolean
  hpo_terms: any[] // Replace with the actual type from your API
  id: string
  inheritance: string
  pseudonym: string
  sex: string
  user: string
  zygosity: string
}

const apiResponseToFrontendCase = (apiResponse: APIResponse): Case => {
  return {
    pseudonym: apiResponse.pseudonym,
    diseases: apiResponse.diseases.map((item) => {
      // Do any conversion required from the API object to OntologyTerm
      return item as OntologyTerm
    }),
    hpoTerms: apiResponse.hpo_terms.map((item) => {
      // Do any conversion required from the API object to OntologyTerm
      return item as OntologyTerm
    }),
    inheritance: apiResponse.inheritance as Inheritance, // Assuming enums can directly map
    affectedFamilyMembers: apiResponse.affected_family_members,
    sex: apiResponse.sex as Sex, // Assuming enums can directly map
    ageOfOnsetMonths: apiResponse.age_of_onset_month,
    ethnicity: apiResponse.ethinicity as Ethnicity, // Assuming enums can directly map
    zygosity: apiResponse.zygosity as Zygosity, // Assuming enums can directly map
    familySegregation: apiResponse.family_segregation
  }
}

export const useCaseStore = defineStore('case', () => {
  /* The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /* The case information. */
  const caseInfo = ref<Case>({
    pseudonym: '',
    diseases: [],
    hpoTerms: [],
    inheritance: Inheritance.Unknown,
    affectedFamilyMembers: null,
    sex: Sex.Unknown,
    ageOfOnsetMonths: null,
    ethnicity: Ethnicity.Unknown,
    zygosity: Zygosity.Unknown,
    familySegregation: null
  })

  function clearData() {
    caseInfo.value = {
      pseudonym: '',
      diseases: [],
      hpoTerms: [],
      inheritance: Inheritance.Unknown,
      affectedFamilyMembers: null,
      sex: Sex.Unknown,
      ageOfOnsetMonths: null,
      ethnicity: Ethnicity.Unknown,
      zygosity: Zygosity.Unknown,
      familySegregation: null
    }
  }

  const loadCase = async () => {
    storeState.value = StoreState.Loading
    try {
      const client = new CaseInfoClient()
      const result: APIResponse = await client.fetchCaseInfo()
      caseInfo.value = apiResponseToFrontendCase(result)
      storeState.value = StoreState.Active
    } catch (e) {
      storeState.value = StoreState.Error
    }
  }

  const updateCase = async (caseUpdate: Case) => {
    caseInfo.value = { ...caseUpdate }
    storeState.value = StoreState.Active
  }

  return {
    storeState,
    caseInfo,
    clearData,
    loadCase,
    updateCase
  }
})
