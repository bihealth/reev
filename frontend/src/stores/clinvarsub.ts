/**
 * Store for manipulating the clinvarsub for the `ClinvarsubCard`.
 *
 * Note that the submitting org editor from the profile page for loading the
 * data as that plays better with the `VDataTableServer`.
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'

import {
  ClinvarsubClient,
  VariantPresence,
  type SubmissionThreadRead,
  type SubmissionThreadWrite,
  type SubmittingOrgRead,
  SubmissionThreadStatus,
  type SubmissionActivityWrite,
  SubmissionActivityStatus,
  SubmissionActivityKind,
  type SubmissionContainer
} from '@/api/clinvarsub'
import { StoreState } from '@/stores/misc'

export const useClinvarsubStore = defineStore('clinvarsub', () => {
  /** The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /** Submitting orgs by UUID. */
  const submittingOrgs = ref<{ [key: string]: SubmittingOrgRead }>({})

  /** The primary variant description that the submission thread etc. are loaded for. */
  const primaryVariantDesc = ref<string | undefined>(undefined)

  /** The `SubmissionThreadRead` objects by UUID for the current primary variant description. */
  const submissionThreads = ref<{ [key: string]: SubmissionThreadRead }>({})

  /** Load all submitting orgs of the current user. */
  const loadSubmittingOrgs = async () => {
    submittingOrgs.value = {}

    const client = new ClinvarsubClient()
    storeState.value = StoreState.Loading
    try {
      let cursorNext: string | undefined = undefined
      do {
        const page = await client.fetchSubmittingOrgs()
        for (const org of page.items) {
          submittingOrgs.value[org.id] = org
        }
        cursorNext = page.next_page ?? undefined
      } while (cursorNext)
      storeState.value = StoreState.Active
    } catch (err) {
      storeState.value = StoreState.Error
    }
  }

  /**
   * Load all submission threads given the primary variant description.
   *
   * Clear by calling with `null`/`undefined`.
   *
   * Will not attempt to load the submision thread for the same primary variant
   * description unless `force` is set to `true`.
   *
   * Will set `storeState` to `StoreState.Error` in case of errors.
   *
   * @param primaryVariantDesc$ The primary variant description to load the submission threads for.
   * @param force Whether to force reloading the submission threads.
   */
  const loadSubmissionThreads = async (
    primaryVariantDesc$: string | null | undefined,
    force: boolean = false
  ) => {
    if (!force && primaryVariantDesc.value === primaryVariantDesc$) {
      return // already loaded
    }

    primaryVariantDesc.value = undefined
    submissionThreads.value = {}
    if (!primaryVariantDesc$) {
      return // cleared
    }

    const client = new ClinvarsubClient()
    storeState.value = StoreState.Loading
    try {
      let cursorNext: string | undefined = undefined
      do {
        const page = await client.fetchSubmissionThreads(primaryVariantDesc$)
        for (const thread of page.items) {
          submissionThreads.value[thread.id] = thread
        }
        cursorNext = page.next_page ?? undefined
      } while (cursorNext)
      primaryVariantDesc.value = primaryVariantDesc$
      storeState.value = StoreState.Active
    } catch (err) {
      storeState.value = StoreState.Error
    }
  }

  /**
   * Create a new submission thread for the current variant via the API.
   *
   * Will update the store state with the newly created submission thread.
   *
   * In case of errors creating, will throw an error but not set the store
   * state.  Will call `loadSubmissionThreads()` that may set the store state
   * to error.
   *
   * @param submittingOrgId The submitting org ID to create the submission thread for.
   * @param primaryVariantDesc$ The primary variant description to create the submission thread for.
   * @param effectiveScv The effective SCV to create the submission thread for.
   * @param effectivePresence The effective presence to create the submission thread for.
   * @param desiredPresence The desired presence to create the submission thread for.
   * @param payload The payload to create the submission thread for.
   * @throws {Error} In case of errors creating the submission thread.
   */
  const createSubmissionThread = async (
    submittingOrgId: string,
    primaryVariantDesc$: string,
    effectiveScv: string | undefined,
    effectivePresence: VariantPresence,
    desiredPresence: VariantPresence,
    payload: SubmissionContainer,
  ): Promise<SubmissionThreadRead> => {
    const client = new ClinvarsubClient()

    // Create submission thread with status 'initial'.
    const submissionThreadCreate: SubmissionThreadWrite = {
      submittingorg_id: submittingOrgId,
      primary_variant_desc: primaryVariantDesc$,
      effective_scv: effectiveScv,
      effective_presence: effectivePresence,
      desired_presence: desiredPresence,
      status: SubmissionThreadStatus.Initial,
    }
    let submissionThread = await client.createSubmissionThread(submissionThreadCreate)
    // Determine activity kind for activity, then create activity.
    let kind: SubmissionActivityKind
    if (desiredPresence === VariantPresence.Absent) {
      kind = SubmissionActivityKind.Delete
    } else if (desiredPresence === VariantPresence.Present) {
      if (effectiveScv === undefined) {
        kind = SubmissionActivityKind.Create
      } else {
        kind = SubmissionActivityKind.Update
      }
    } else {
      throw new Error(`unexpected kind: ${desiredPresence}/${effectiveScv}`)
    }
    const submissionActivityCreate: SubmissionActivityWrite = {
      submissionthread_id: submissionThread.id,
      kind,
      status: SubmissionActivityStatus.Initial,
      request_payload: payload,
    }
    let submissionActivity = await client.createSubmissionActivity(submissionActivityCreate)
    // Now that thread/activity are on the server, submit both to the worker by
    // marking them as `WAITING`.
    const submissionThreadUpdate: SubmissionThreadWrite = {
      ...submissionThread,
      status: SubmissionThreadStatus.Waiting,
    }
    submissionThread = await client.updateSubmissionThread(submissionThreadUpdate)
    const submissionActivityUpdate: SubmissionActivityWrite = {
      ...submissionActivity,
      status: SubmissionActivityStatus.Waiting,
    }
    submissionActivity = await client.updateSubmissionActivity(submissionActivityUpdate)

    // Finally, reload submission threads.
    loadSubmissionThreads(primaryVariantDesc$, true)

    return submissionThread
  }

  /**
   * Initialize the store.
   *
   * When forced, this will reload all data.
   */
  const initialize = async (force: boolean = false) => {
    // guard against initializing multiple times
    if (storeState.value !== StoreState.Initial && !force) {
      return
    }
    // clear against artifacts, mark store as currently loading
    clear()
    storeState.value = StoreState.Loading
    // load all submitting orgs
    await loadSubmittingOrgs()
    // initialization done; mark store as active
    storeState.value = StoreState.Active
  }

  /**
   * Clear all data in the store.
   */
  const clear = () => {
    storeState.value = StoreState.Initial
    submittingOrgs.value = {}
    primaryVariantDesc.value = undefined
    submissionThreads.value = {}
  }

  return {
    storeState,
    submittingOrgs,
    primaryVariantDesc,
    submissionThreads,
    clear,
    initialize,
    loadSubmittingOrgs,
    loadSubmissionThreads,
    createSubmissionThread
  }
})
