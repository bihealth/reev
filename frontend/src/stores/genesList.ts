/** Store for gene details.
 *
 * This includes the data retrieved from the APIs.
 */

import { defineStore } from 'pinia'
import { ref } from 'vue'

import { StoreState } from '@/stores/misc'
import { AnnonarsClient } from '@/api/annonars'

export const useGenesListStore = defineStore('genesList', () => {
  /* The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /* The current search query. */
  const query = ref<string | null>(null)

  /* The retrieved data for search query. */
  const genesList = ref<any | null>(null)

  /* The redirect HGNC ID. Is used only if storeState is "Redirect".*/
  const redirectHgncId = ref<string | null>(null)

  function clearData() {
    storeState.value = StoreState.Initial
    query.value = null
    genesList.value = null
  }

  const checkResultsMatch = async () => {
    if (query.value === null) {
      return null
    }
    const queryArray = query.value.split('&')
    const searchTerm = queryArray[0].split('=')[1]

    for (const gene of genesList.value) {
      if (gene.data.hgnc_id === searchTerm) {
        return gene.data.hgnc_id
      }
      if (gene.data.symbol === searchTerm) {
        return gene.data.hgnc_id
      }
      if (gene.data.ensembl_gene_id === searchTerm) {
        return gene.data.hgnc_id
      }
      if (gene.data.ncbi_gene_id === searchTerm) {
        return gene.data.hgnc_id
      }
    }
    return null
  }

  const loadData = async (geneSymbolQuery: Object) => {
    // Form string from object
    const geneQuery: string = Object.entries(geneSymbolQuery)
      .map(([key, val]) => `${key}=${val}`)
      .join('&')

    // Do not re-load data if the gene symbol is the same
    if (geneQuery === query.value) {
      return
    }

    // Clear against artifact
    clearData()

    // Load data via API
    storeState.value = StoreState.Loading
    try {
      const client = new AnnonarsClient()
      const data = await client.fetchGenes(geneQuery)
      if (data.genes.length === 0) {
        throw new Error('No data returned from API.')
      }
      genesList.value = data['genes']

      query.value = geneQuery
      const hgncId = await checkResultsMatch()
      if (hgncId !== null) {
        storeState.value = StoreState.Redirect
        redirectHgncId.value = hgncId
      } else {
        storeState.value = StoreState.Active
      }
    } catch (e) {
      console.error('There was an error while searching for genes.', e)
      storeState.value = StoreState.Error
    }
  }

  return {
    storeState,
    query,
    genesList,
    redirectHgncId,
    loadData,
    clearData
  }
})
