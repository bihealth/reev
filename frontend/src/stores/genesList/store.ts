/**
 * Store for gene details.
 *
 * This includes the data retrieved from the APIs.
 */
import { AnnonarsClient, ScoreGeneNames } from '@bihealth/reev-frontend-lib/api/annonars'
import { StoreState } from '@bihealth/reev-frontend-lib/stores'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useGenesListStore = defineStore('genesList', () => {
  /* The current store state. */
  const storeState = ref<StoreState>(StoreState.Initial)

  /* The current search query. */
  const query = ref<string | undefined>(undefined)

  /* The retrieved data for search query. */
  const genesList = ref<ScoreGeneNames[] | undefined>(undefined)

  /* The redirect HGNC ID. Is used only if storeState is "Redirect".*/
  const redirectHgncId = ref<string | undefined>(undefined)

  function clearData() {
    storeState.value = StoreState.Initial
    query.value = undefined
    genesList.value = undefined
  }

  const checkResultsMatch = async () => {
    if (query.value === undefined || genesList.value === undefined) {
      return undefined
    }
    const queryArray = query.value.split('&')
    const searchTerm = queryArray[0].split('=')[1]
    for (const gene of genesList.value) {
      if (gene.data.hgncId === searchTerm) {
        return gene.data.hgncId
      }
      if (gene.data.symbol === searchTerm) {
        return gene.data.hgncId
      }
      if (gene.data.ensemblGeneId === searchTerm) {
        return gene.data.hgncId
      }
      if (gene.data.ncbiGeneId === searchTerm) {
        return gene.data.hgncId
      }
    }
    return undefined
  }

  const loadData = async (geneSymbolQuery: object) => {
    // Form string from object
    const geneQuery: string = Object.entries(geneSymbolQuery)
      .map(([key, val]) => `${key}=${val}`)
      .join('&')

    // Do not re-load data if the gene symbol is the same
    if (geneQuery === query.value) {
      return
    }

    // Clear against artifact
    clearData()

    // Load data via API
    storeState.value = StoreState.Loading
    try {
      const client = new AnnonarsClient()
      const response = await client.fetchGenes(geneQuery)
      if (response.genes.length === 0) {
        throw new Error('No data returned from API.')
      }
      genesList.value = response.genes

      query.value = geneQuery
      const hgncId = await checkResultsMatch()
      if (hgncId !== undefined) {
        storeState.value = StoreState.Redirect
        redirectHgncId.value = hgncId
      } else {
        storeState.value = StoreState.Active
      }
    } catch (e) {
      console.error('There was an error while searching for genes.', e)
      clearData()
      storeState.value = StoreState.Error
    }
  }

  return {
    storeState,
    query,
    genesList,
    redirectHgncId,
    loadData,
    clearData
  }
})
