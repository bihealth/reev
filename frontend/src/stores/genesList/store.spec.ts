import { StoreState } from '@bihealth/reev-frontend-lib/stores'
import { createPinia, setActivePinia } from 'pinia'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

import { useGenesListStore } from '../genesList'

const fetchMocker = createFetchMock(vi)

/** Example list of genes */
const exampleGenesList = {
  genes: [
    {
      score: 0.75,
      data: {
        hgnc_id: 'HGNC:3333',
        symbol: 'EMP1',
        name: 'epithelial membrane protein 1',
        alias_symbol: ['TMP', 'CL-20'],
        alias_name: [],
        ensembl_gene_id: 'ENSG00000134531',
        ncbi_gene_id: '2012'
      }
    },
    {
      score: 0.75,
      data: {
        hgnc_id: 'HGNC:3334',
        symbol: 'EMP2'
      }
    }
  ]
}

describe('geneInfo Store', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    fetchMocker.enableMocks()
    fetchMocker.resetMocks()
  })

  it('should have initial state', () => {
    // arrange:
    const store = useGenesListStore()

    // act: nothing to do

    // assert:
    expect(store.storeState).toBe(StoreState.Initial)
    expect(store.query).toBe(undefined)
    expect(store.genesList).toBe(undefined)
    expect(store.redirectHgncId).toBe(undefined)
  })

  it('should clear state', () => {
    // arrange:
    const store = useGenesListStore()
    store.storeState = StoreState.Active
    store.query = 'q=BRCA1&fields=symbol'
    store.genesList = JSON.parse(JSON.stringify({ gene: 'info' }))

    // act:
    store.clearData()

    // assert:
    expect(store.storeState).toBe(StoreState.Initial)
    expect(store.query).toBe(undefined)
    expect(store.genesList).toBe(undefined)
  })

  it('should load data', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'EMP', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Active)
    expect(store.query).toBe('q=EMP&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).matchSnapshot()
  })

  it('should fail to load data with invalid request', async () => {
    // arrange:
    // Disable error logging
    vi.spyOn(console, 'error').mockImplementation(() => {})
    fetchMocker.mockResponseOnce(JSON.stringify({ foo: 'bar' }), { status: 400 })
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'XXX', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Error)
    expect(store.query).toBe(undefined)
    expect(store.genesList).toBe(undefined)
  })

  it('should fail to load data with invalid fetchGenes response', async () => {
    // arrange:
    // Disable error logging
    vi.spyOn(console, 'error').mockImplementation(() => {})
    fetchMocker.mockResponseOnce(JSON.stringify({ genes: [] }))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'EMP', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(console.error).toHaveBeenCalled()
    expect(console.error).toHaveBeenCalledWith(
      'There was an error while searching for genes.',
      new Error('No data returned from API.')
    )
    expect(store.storeState).toBe(StoreState.Error)
    expect(store.query).toBe(undefined)
    expect(store.genesList).toBe(undefined)
  })

  it('should not load data if gene symbol is the same', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'EMP', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Active)
    expect(store.query).toBe('q=EMP&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).matchSnapshot()

    // act2:
    await store.loadData({ q: 'EMP', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert2:
    expect(fetchMocker.mock.calls.length).toBe(1)
  })

  it('should redirect if the searchTerm matched symbol', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'EMP1', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Redirect)
    expect(store.redirectHgncId).toBe('HGNC:3333')
    expect(store.query).toBe('q=EMP1&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).matchSnapshot()
    expect(fetchMocker.mock.calls.length).toBe(1)
  })

  it('should redirect if the searchTerm matched hgnc_id', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: 'HGNC:3333', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Redirect)
    expect(store.redirectHgncId).toBe('HGNC:3333')
    expect(store.query).toBe('q=HGNC:3333&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).matchSnapshot()
    expect(fetchMocker.mock.calls.length).toBe(1)
  })

  it('should redirect if the searchTerm matched ensembl_gene_id', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({
      q: 'ENSG00000134531',
      fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol'
    })

    // assert:
    expect(store.storeState).toBe(StoreState.Redirect)
    expect(store.redirectHgncId).toBe('HGNC:3333')
    expect(store.query).toBe('q=ENSG00000134531&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).matchSnapshot()
    expect(fetchMocker.mock.calls.length).toBe(1)
  })

  it('should redirect if the searchTerm matched ncbi_gene_id', async () => {
    // arrange:
    fetchMocker.mockResponse(JSON.stringify(exampleGenesList))
    const store = useGenesListStore()

    // act:
    await store.loadData({ q: '2012', fields: 'hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol' })

    // assert:
    expect(store.storeState).toBe(StoreState.Redirect)
    expect(store.redirectHgncId).toBe('HGNC:3333')
    expect(store.query).toBe('q=2012&fields=hgnc_id,ensembl_gene_id,ncbi_gene_id,symbol')
    expect(store.genesList).toMatchSnapshot()
    expect(fetchMocker.mock.calls.length).toBe(1)
  })
})
