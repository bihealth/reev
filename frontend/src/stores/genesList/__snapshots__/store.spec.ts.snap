// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`geneInfo Store > should load data 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;

exports[`geneInfo Store > should not load data if gene symbol is the same 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;

exports[`geneInfo Store > should redirect if the searchTerm matched ensembl_gene_id 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;

exports[`geneInfo Store > should redirect if the searchTerm matched hgnc_id 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;

exports[`geneInfo Store > should redirect if the searchTerm matched ncbi_gene_id 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;

exports[`geneInfo Store > should redirect if the searchTerm matched symbol 1`] = `
[
  {
    "data": {
      "aliasName": [],
      "aliasSymbol": [
        "TMP",
        "CL-20",
      ],
      "ensemblGeneId": "ENSG00000134531",
      "hgncId": "HGNC:3333",
      "name": "epithelial membrane protein 1",
      "ncbiGeneId": "2012",
      "symbol": "EMP1",
    },
    "score": 0.75,
  },
  {
    "data": {
      "aliasName": undefined,
      "aliasSymbol": undefined,
      "ensemblGeneId": undefined,
      "hgncId": "HGNC:3334",
      "name": undefined,
      "ncbiGeneId": undefined,
      "symbol": "EMP2",
    },
    "score": 0.75,
  },
]
`;
